File: OpenMP_OpenACC/F/openacc_iccg/1_test/solver_ICCG_mc.f
!** 
!C*** module solver_ICCG_mc
!C***
!
      module solver_ICCG_mc
      contains
!C
!C*** solve_ICCG
!C
      subroutine solve_ICCG_mc                                          &
     &         ( N, NPL, NPU, indexL, itemL, indexU, itemU, D, B, X,    &
     &           AL, AU, NCOLORtot, PEsmpTOT, SMPindex, SMPindexG,      &
     &           EPS, ITR, IER)

      use omp_lib
      implicit REAL*8 (A-H,O-Z)

      integer :: N, NL, NU, NCOLORtot, PEsmpTOT

      real(kind=8), dimension(N)   :: D
      real(kind=8), dimension(N)   :: B
      real(kind=8), dimension(N)   :: X

      real(kind=8), dimension(NPL) :: AL
      real(kind=8), dimension(NPU) :: AU

      integer, dimension(0:N)   :: indexL, indexU
      integer, dimension(NPL):: itemL
      integer, dimension(NPU):: itemU

      integer, dimension(0:NCOLORtot*PEsmpTOT):: SMPindex
      integer, dimension(0:PEsmpTOT)       :: SMPindexG

      real(kind=8), dimension(:,:), allocatable :: W

      integer, parameter ::  R= 1
      integer, parameter ::  Z= 2
      integer, parameter ::  Q= 2
      integer, parameter ::  P= 3
      integer, parameter :: DD= 4

!C
!C +------+
!C | INIT |
!C +------+
!C===

      allocate (W(N,4))

!$acc enter data
!$acc& copyin(B, D, indexL, itemL)
!$acc& copyin(indexU, itemU, AL, AU)
!$acc& create(X, W)



!$omp parallel do private(i)
!$acc kernels present(X, W, B)
!$acc loop independent
      do i= 1, N
        X(i)  = 0.d0
        W(i,2)= 0.0D0
        W(i,3)= 0.0D0
        W(i,4)= 0.0D0
        W(i,R)= B(i)
      enddo
!$acc end kernels

!$omp parallel private(ic,ip1,ip2,i,VAL,k)
      do ic= 1, NCOLORtot
        ip1= SMPindex((ic-1)*PEsmpTOT) + 1
        ip2= SMPindex((ic-1)*PEsmpTOT + PEsmpTOT)
!$omp do
!$acc kernels present(D, AL, indexL, itemL, W) 
!$acc loop independent
        do i= ip1, ip2
          VAL= D(i)
          !$acc loop seq
          do k= indexL(i-1)+1, indexL(i)
            VAL= VAL - (AL(k)**2) * W(itemL(k),DD)
          enddo
          W(i,DD)= 1.d0/VAL
        enddo
!$acc end kernels
      enddo
!$omp end parallel

!$acc exit data
!$acc& delete(B, D, indexL, itemL)
!$acc& delete(indexU, itemU, AL, AU, W)
!$acc& copyout(X)
!C===

!C
!C +-----------------------+
!C | {r0}= {b} - [A]{xini} |
!C +-----------------------+
!C===

      BNRM2= 0.0D0
!$omp parallel do private(i) reduction(+:BNRM2)
      do i = 1, N
        BNRM2 = BNRM2 + B(i)  **2
      enddo
!C===

!C
!C***************************************************************  ITERATION
      ITR= N

      do L= 1, ITR
!C
!C +----------------+
!C | {z}= [Minv]{r} |
!C +----------------+
!C===      

!$omp parallel do private(i)
      do i = 1, N
        W(i,Z)= W(i,R)
      enddo


!$omp parallel private(ic,ip1,ip2,i,WVAL,k)
      do ic= 1, NCOLORtot
        ip1= SMPindex((ic-1)*PEsmpTOT) + 1
        ip2= SMPindex((ic-1)*PEsmpTOT + PEsmpTOT)
!$omp do
        do i= ip1, ip2
          WVAL= W(i,Z)
          do k= indexL(i-1)+1, indexL(i)
            WVAL= WVAL -  AL(k) * W(itemL(k),Z)
          enddo
          W(i,Z)= WVAL * W(i,DD)
        enddo
      enddo
!$omp end parallel


!$omp parallel private(ic,ip1,ip2,i,SW,k)
      do ic= NCOLORtot, 1, -1
        ip1= SMPindex((ic-1)*PEsmpTOT) + 1
        ip2= SMPindex((ic-1)*PEsmpTOT + PEsmpTOT)
!$omp do
        do i= ip1, ip2
          SW  = 0.0d0
          do k= indexU(i-1)+1, indexU(i)
            SW= SW + AU(k) * W(itemU(k),Z)
          enddo
          W(i,Z)= W(i,Z) - W(i,DD) * SW
        enddo
      enddo
!$omp end parallel
!C===

!C
!C +-------------+
!C | RHO= {r}{z} |
!C +-------------+
!C===
      RHO= 0.d0
!$omp parallel do private(i) reduction(+:RHO)
      do i = 1, N
        RHO= RHO + W(i,R)*W(i,Z)   
      enddo
!C===

!C
!C +-----------------------------+
!C | {p} = {z} if      ITER=1    |
!C | BETA= RHO / RHO1  otherwise |
!C +-----------------------------+
!C===
      if ( L.eq.1 ) then
!$omp parallel do private(i)
        do i = 1, N
          W(i,P)= W(i,Z)
        enddo
       else
        BETA= RHO / RHO1
!$omp parallel do private(i)
        do i = 1, N
          W(i,P)= W(i,Z) + BETA*W(i,P)
        enddo
      endif
!C===

!C
!C +-------------+
!C | {q}= [A]{p} |
!C +-------------+
!C===        
!$omp parallel do private(i,VAL,k)
      do i= 1, N
        VAL= D(i)*W(i,P)
        do k= indexL(i-1)+1, indexL(i)
          VAL= VAL + AL(k)*W(itemL(k),P)
        enddo 
        do k= indexU(i-1)+1, indexU(i)
          VAL= VAL + AU(k)*W(itemU(k),P)
        enddo 
        W(i,Q)= VAL
      enddo
!C===

!C
!C +---------------------+
!C | ALPHA= RHO / {p}{q} |
!C +---------------------+
!C===
      C1= 0.d0
!$omp parallel do private(i) reduction(+:C1)
      do i = 1, N
        C1= C1 + W(i,P)*W(i,Q)
      enddo

      ALPHA= RHO / C1
!C===


!C
!C +----------------------+
!C | {x}= {x} + ALPHA*{p} |
!C | {r}= {r} - ALPHA*{q} |
!C +----------------------+
!C===
!$omp parallel do private(i)
      do i = 1, N
        X(i)  = X(i)   + ALPHA * W(i,P)
        W(i,R)= W(i,R) - ALPHA * W(i,Q)
      enddo

      DNRM2= 0.d0
!$omp parallel do private(i) reduction(+:DNRM2)
      do i = 1, N
        DNRM2= DNRM2 + W(i,R)**2
      enddo
!C===

      ERR = dsqrt(DNRM2/BNRM2)
      if (mod(L,100).eq.1) then
        write (*,'(i5,2(1pe16.6))') L, ERR
      endif

        if (ERR .lt. EPS) then
          IER = 0
          goto 900
         else
          RHO1 = RHO
        endif

      enddo
      IER = 1

  900 continue

      write (*,'(i5,2(1pe16.6))') L, ERR
      ITR= L
!      EPS= ERR

      deallocate (W)

      return

      end subroutine  solve_ICCG_mc
      end module     solver_ICCG_mc

File: OpenMP_OpenACC/C/openacc_iccg/06_optimized/solver_ICCG_mc.c
/*
 * solver_ICCG
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <math.h>

#include "solver_ICCG_mc.h"

extern int
solve_ICCG_mc(int N, int NL, int NU, int NPL, int NPU, int *indexL, int *itemL, int *indexU, int *itemU,
	      double *D, double *B, double *X, double *AL, double *AU,
	      int NCOLORtot, int PEsmpTOT, int *SMPindex, int *SMPindexG,
	      double EPS, int *ITR, int *IER)
{
        double **W;
        double VAL, BNRM2, WVAL, SW, RHO, BETA, RHO1, C1, DNRM2, ALPHA, ERR;
	int i, j, ic, ip, L, ip1, ip2;
	int R = 0;
	int Z = 1;
	int Q = 1;
	int P = 2;
	int DD = 3;

/*********
 * INIT. *
 *********/

	//	W = (double **)allocate_matrix(sizeof(double), 4, N);
	allocate_matrix2(&W,sizeof(double), 4, N);

#pragma acc enter data                  \
  copyin(D[0:N])                                \
  copyin(X[0:N])                                \
  copyin(indexL[0:N+1])                         \
  copyin(AL[0:NPL])                             \
  copyin(itemL[0:NPL])                          \
  copyin(indexU[0:N+1])                         \
  copyin(AU[0:NPU])                             \
  copyin(itemU[0:NPU])                          \
  copyin(B[0:N])                                \
  copyin(SMPindex[0:NCOLORtot*PEsmpTOT])        \
  create(W[0:4][0:N])
  

#pragma omp parallel for private (i)
#pragma acc kernels present(X,W)
#pragma acc loop independent
	  for(i=0; i<N; i++) {
	    X[i] = 0.0;
	    W[1][i] = 0.0;
	    W[2][i] = 0.0;
	    W[3][i] = 0.0;
	  }
	  
#pragma omp parallel private (ic, ip1, ip2, i, WVAL, j)
	  for(ic=0; ic<NCOLORtot; ic++) {
	    ip1 = ic * PEsmpTOT;
	    ip2 = ic * PEsmpTOT + PEsmpTOT;
#pragma omp for
#pragma acc kernels present(SMPindex,D,indexL,AL,W,itemL,W)
#pragma acc loop independent
	    for(i=SMPindex[ip1]; i<SMPindex[ip2]; i++) {
	      VAL = D[i];
	      for(j=indexL[i]; j<indexL[i+1]; j++) {
		VAL = VAL - AL[j]*AL[j] * W[DD][itemL[j] - 1];
	      }
	      W[DD][i] = 1.0 / VAL;
	    }
	  }
	  
	  /**************************
	   * {r0} = {b} - {A}{xini} *
	   **************************/
#pragma omp parallel for private (i, VAL, j)
#pragma acc kernels present(D,X,indexL,AL,itemL,indexU,AU,itemU,W,B)
#pragma acc loop independent
	  for(i=0; i<N; i++) {
	    VAL = D[i] * X[i];
	    for(j=indexL[i]; j<indexL[i+1]; j++) {
	      VAL += AL[j] * X[itemL[j]-1];
	    }
	    for(j=indexU[i]; j<indexU[i+1]; j++) {
	      VAL += AU[j] * X[itemU[j]-1];
	    }
	    W[R][i] = B[i] - VAL;
	  }
	  
	  BNRM2 = 0.0;
#pragma omp parallel for private (i) reduction (+:BNRM2)
#pragma acc kernels  present(B)
#pragma acc loop reduction (+:BNRM2)
	  for(i=0; i<N; i++) {
	    BNRM2 += B[i]*B[i];
	  }
	  
	  /************************************************************** ITERATION */
	  *ITR = N;
	  
	  for(L=0; L<(*ITR); L++) {
	    
	    /*******************
	     * {z} = [Minv]{r} *
	     *******************/
#pragma omp parallel for private(i)
#pragma acc kernels present(W) async(0)
#pragma acc loop independent
	    for(i=0; i<N; i++) {
	      W[Z][i] = W[R][i];
	    }
	    
#pragma omp parallel private (ic, ip1, ip2, i, WVAL, j)
	    for(ic=0; ic<NCOLORtot; ic++) {
	      ip1 = ic * PEsmpTOT;
	      ip2 = ic * PEsmpTOT + PEsmpTOT;
#pragma omp for
#pragma acc kernels present(SMPindex,indexL,AL,itemL,W) async(0)
#pragma acc loop independent
	      for(i=SMPindex[ip1]; i<SMPindex[ip2]; i++) {
		WVAL = W[Z][i];
		for(j=indexL[i]; j<indexL[i+1]; j++) {
		  WVAL -= AL[j] * W[Z][itemL[j]-1];
		}
		W[Z][i] = WVAL * W[DD][i];
	      }
	    }
	    
#pragma omp parallel private (ic,ip1, ip2, i, SW, j)
	    for(ic=NCOLORtot-1; ic>=0; ic--) {
	      ip1 = ic * PEsmpTOT;
	      ip2 = ic * PEsmpTOT + PEsmpTOT;
#pragma omp for
#pragma acc kernels present(SMPindex,indexU,AU,itemU,W) async(0)
#pragma acc loop independent
	      for(i=SMPindex[ip1]; i<SMPindex[ip2]; i++) {
		SW = 0.0;
		for(j=indexU[i]; j<indexU[i+1]; j++) {
		  SW += AU[j] * W[Z][itemU[j]-1];
		}
		W[Z][i] -= W[DD][i] * SW;
	      }
	    }
	    
	    /****************
	     * RHO = {r}{z} *
	     ****************/
	    RHO = 0.0;
#pragma omp parallel for private (i) reduction (+:RHO)
#pragma acc kernels present(W) async(0)
#pragma acc loop reduction(+:RHO) gang vector(256)
	    for(i=0; i<N; i++) {
	      RHO += W[R][i] * W[Z][i];
	    }
#pragma acc wait(0)
	    
	    /********************************
	     * {p}  = {z} if      ITER=0    *
	     * BETA = RHO / RHO1  otherwise *
	     ********************************/
	    if(L == 0) {
#pragma omp parallel for private (i)
#pragma acc kernels present(W) async(0)
#pragma acc loop independent
	      for(i=0; i<N; i++) {
		W[P][i] = W[Z][i];
	      }
	    } else {
	      BETA = RHO / RHO1;
#pragma omp parallel for private (i)
#pragma acc kernels present(W) async(0)
#pragma acc loop independent
	      for(i=0; i<N; i++) {
		W[P][i] = W[Z][i] + BETA * W[P][i];
	      }
	    }
	    
	    /****************
	     * {q} = [A]{p} *
	     ****************/
#pragma omp parallel for private (i, VAL, j)
#pragma acc kernels present(D,W,indexL,AL,itemL,indexU,AU,itemU)  async(0)
#pragma acc loop independent
	    for(i=0; i<N; i++) {
	      VAL = D[i] * W[P][i];
	      for(j=indexL[i]; j<indexL[i+1]; j++) {
		VAL += AL[j] * W[P][itemL[j]-1];
	      }
	      for(j=indexU[i]; j<indexU[i+1]; j++) {
		VAL += AU[j] * W[P][itemU[j]-1];
	      }
	      W[Q][i] = VAL;
	    }
	    
	    /************************
	     * ALPHA = RHO / {p}{q} *
	     ************************/
	    C1 = 0.0;
#pragma omp parallel for private (i) reduction (+:C1)
#pragma acc kernels present(W)  async(0)
#pragma acc loop reduction(+:C1) gang vector(256)
	    for(i=0; i<N; i++) {
	      C1 += W[P][i] * W[Q][i];
	    }
#pragma acc wait(0)
	    
	    ALPHA = RHO / C1;
	    
	    /***************************
	     * {x} = {x} + ALPHA * {p} *
	     * {r} = {r} - ALPHA * {q} *
	     ***************************/
#pragma omp parallel for private (i)
#pragma acc kernels present(X,W)  async(0)
#pragma acc loop independent
	    for(i=0; i<N; i++) {
	      X[i]    += ALPHA * W[P][i];
	    }
	    
#pragma omp parallel for private (i)
#pragma acc kernels present(W) async(0)
#pragma acc loop independent
	    for(i=0; i<N; i++) {
	      W[R][i] -= ALPHA * W[Q][i];
	    }
	    
	    DNRM2 = 0.0;
#pragma omp parallel for private (i) reduction (+:DNRM2)
#pragma acc kernels present(W) async(0)
#pragma acc loop reduction(+:DNRM2) gang vector(256)
	    for(i=0; i<N; i++) {
	      DNRM2 += W[R][i]*W[R][i];
	    }
#pragma acc wait(0)
	    
	    ERR = sqrt(DNRM2/BNRM2);
	    if( (L+1)%100 ==1) {
	      fprintf(stderr, "%5d%16.6e\n", L+1, ERR);
	    }
	    
	    if(ERR < EPS) {
	      *IER = 0;
	      goto N900;
	    } else {
	      RHO1 = RHO;
	    }
	  }
	  *IER = 1;
	N900:

#pragma acc exit data                           \
  delete(D[0:N])                                \
  copyout(X[0:N])                               \
  delete(indexL[0:N+1])                         \
  delete(AL[0:NPL])                             \
  delete(itemL[0:NPL])                          \
  delete(indexU[0:N+1])                         \
  delete(AU[0:NPU])                             \
  delete(itemU[0:NPU])                          \
  delete(B[0:N])                                \
  delete(SMPindex[0:NCOLORtot*PEsmpTOT])        \
  delete(W[0:4][0:N])

	fprintf(stderr, "%5d%16.6e\n", L+1, ERR);
	*ITR = L;
	
	free(W);
	
	return 0;
}
